src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.OpenFileTracker.cs

```diff
@@ -165,20 +76,20 @@ private void TryOpeningDocumentsForNewCookie(uint cookie)
                     }
                     else
                     {
-                        activeContextProjectId = GetActiveContextProjectIdAndWatchHierarchies(cookie, documentIds.Select(d => d.ProjectId));
+                        activeContextProjectId = GetActiveContextProjectId(cookie, documentIds.Select(d => d.ProjectId));
                     }
 
-                    if ((object)_runningDocumentTable.GetDocumentData(cookie) is IVsTextBuffer bufferAdapter)
+                    if ((object)RunningDocumentTable.GetDocumentData(cookie) is IVsTextBuffer bufferAdapter)
```

Comment:
My expectation would have been that all of this still lives in the base. The only bit that would be in the derived methods would be the stuff that starts on `if (w.CurrentSolution.ContainsDocument(…))` stuff, where the input to the function is a DocumentId and the SourceTextContainer.

This is of course tricky  because of the locking semantics, so you might have to have two methods that are implemented by derivers

    void ApplyWorkspaceChangeUnderLock(Action<Workspace> w);

which just forwards to the VS workspace one that acquires the lock, and then:

    void ApplyDocumentOpen(DocumentId, SourceTextContainer)